GFDLMP-v3 and TEMPO MP microphysics

dustinswales opened the pull request:
## Description (*updated on 7/2/2025)

This PR contains CCPP implementations of two cloud microphysics schemes (see below).

**_GFDLMP v3:_**
@RuiyuSun 

This PR is intended to supersede #162

For the refactoring, all of the configuration/namelist info is moved into a common module that is referenced by both MP versions and a shared fv_sat_adj.F90. There are default values for the scheme configurations defined within this module, and overwritten by the namelist, if provided. Both the v1 and v3 rely on some of the same parameters, but with different default values. This is handled by introducing a pre-processor directive into the new configuration module.

I also removed the imp_physics_gfdl_v3 flag. GFDL v1 and v3 have their own distinct namelists, so there is no need to have another switch to distinguish v1/v3.
The only differences to run with GFDLv1 or v3 are:
- in the SDF: gfdl_cloud_microphys or gfdl_cloud_microphys_v3 and
- which namelist is added to gfs_phys_nml: [V1 nml](https://github.com/dustinswales/ccpp-physics/blob/feature/gfdlmpv3/physics/MP/GFDL/v1_2019/module_gfdl_cloud_microphys.F90#L323) or [V3 nml](https://github.com/dustinswales/ccpp-physics/blob/feature/gfdlmpv3/physics/MP/GFDL/v3_2022/module_gfdl_cloud_microphys_v3.F90#L529)
- set new pre-processor directive, GFDLMP_v3, to true.

**_TEMPO MP:_**
@AndersJensen-NOAA 

This PR replaces #245 
Development of the Thompson-Eidhammer microphysics is transitioning to a submodule, and this submodule is: https://github.com/NCAR/TEMPO

TEMPO stand for Thompson-Eidhammer Microphysics Parameterization for Operations.

The use of a submodule will enable centralized community development (and testing) of the microphysics scheme for various applications and dynamical cores. This submodule will eventually replace the Thompson parameterization currently in the CCPP after approval from all parties involved and significant testing.


## Testing 
Tested on Hera using Intel/GNU
No changes to baselines

RuiyuSun committed:
deleted gfdl mp v1 files

RuiyuSun committed:
removed the gfdlmp v1 files in the GFDL directory and moved them to v1 subdirectory

RuiyuSun committed:
make a change in CMakeLists.txt corresponging to the moving of GFLDMP into a subdir v1

RuiyuSun committed:
removed GFDL directory but added GFDL_2019_v1 for GFDLMP v1 and moved the associated files

RuiyuSun committed:
re-organized GFDLMP v1 and add v3

RuiyuSun committed:
add GFDLMP v3

RuiyuSun committed:
remove comment and previous tests

RuiyuSun committed:
remove previous comments

RuiyuSun committed:
remove comments in module_gfdl_cld_mp.F90

RuiyuSun committed:
correct a bug

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into HEAD

dustinswales committed:
refactor gfdlmp. work in progress

dustinswales committed:
Some more changes

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Additional cleanup

dustinswales committed:
Revert some changes

dustinswales committed:
Revert some more changes

dustinswales committed:
Revert some even more changes

RuiyuSun commented:
@dustinswales This is great and thanks!! I agree that the scheme name in SDF and different namelist name are enough to distinguish the two versions of the scheme. So I think it is good idea to remove the  imp_physics_gfdl_v3 flag. Did you forget to remove the flag in the gfdl_cloud_microphysics_v3.F90 file? 

What is next?

dustinswales committed:
Remove reference to gfdlmp_v3 switch in scheme file

dustinswales commented:
@RuiyuSun Good call. I just removed the flag from the scheme.

Next we need to test that the RTs that use GFDLMP v1 are unchanged (I can do this next week). I can't speak to how the v3 results should differ wrt v1 (I will leave this piece to you)

RuiyuSun commented:
> @RuiyuSun Good call. I just removed the flag from the scheme.
> 
> Next we need to test that the RTs that use GFDLMP v1 are unchanged (I can do this next week). I can't speak to how the v3 results should differ wrt v1 (I will leave this piece to you)

@dustinswales Sounds good.

RuiyuSun commented:
@dustinswales How do I check out this code? Thanks!

dustinswales commented:
> @dustinswales How do I check out this code? Thanks!

git clone --recursive --branch feature/gfdlmpv3 https://github.com/dustinswales/ufs-weather-model.git

dustinswales committed:
Some housekeeping

dustinswales committed:
Some more changes

dustinswales committed:
Some reorg

grantfirl commented:
@dustinswales @RuiyuSun This isn't yet ready to be put in the merge queue, correct?

grantfirl commented:
@dustinswales @RuiyuSun Does this supercede #162 so that it can be closed?

dustinswales commented:
@grantfirl Yes. #162 can be closed.
More changes coming to this PR in the coming days.

dustinswales committed:
Some more cleanup. Nearly done

grantfirl commented:
@dustinswales @RuiyuSun Will you let us know when this is ready to review and test?

dustinswales committed:
Move gfdlmp init call to fv_sat_adj init().

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Revert "Move gfdlmp init call to fv_sat_adj init()."

This reverts commit 3115d2ce905d0b0c6400c7215606586558b1416d.

dustinswales committed:
Revert mp init move.

dustinswales committed:
Some more changes.

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Some more changes.

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Some of the last changes

dustinswales committed:
More changes!

dustinswales committed:
Final updates

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Cleanup module

dustinswales committed:
More changes...

dustinswales committed:
Bug fix

dustinswales committed:
Thats it!

dustinswales commented:
@grantfirl This is ready for review.

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Bug fix from GNU compile

dustinswales committed:
Doh!

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  real(kind_phys) :: ql_mlt         = 2.0e-3  !< (v1/v3) max value of cloud water allowed from melted cloud ice
+  real(kind_phys) :: qs_mlt         = 1.0e-6  !< (v1/v3) max cloud water due to snow melt
+  real(kind_phys) :: ql_gen         = 1.0e-3  !< (v1/v3) max cloud water generation during remapping step if fast_sat_adj = .t.
+  real(kind_phys) :: qi_gen         = 1.82e-6 !< (v1/v3) max cloud ice generation during remapping step (V1 ONLY. Computed internally in V3)
```
```suggestion
  real(kind_phys) :: qi_gen         = 1.82e-6 !< max cloud ice generation during remapping step (V1 ONLY. Computed internally in V3)
```

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  ! #####################################################################################
+  ! GFDL MP Version 3 parameters
+  ! #####################################################################################
+  logical :: const_vw               = .false. !< if .ture., the constants are specified by v * _fac
```
```suggestion
  logical :: const_vw               = .false. !< if .true., the constants are specified by v * _fac
```

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  integer :: radr_flag              = 1       !< radar reflectivity for rain
+                                              !< 1: Mark Stoelinga (2005)
+                                              !< 2: Smith et al. (1975), Tong and Xue (2005)
+                                              !< 3: Marshall-Palmer formula (https://en.wikipedia.org/wiki/DBZ_(meteorology))
```
More authoritative source. Could also just state "Marshall and Palmer 1948"
```suggestion
                                              !< 3: Marshall-Palmer formula (https://glossary.ametsoc.org/wiki/Marshall-palmer_relation)
```

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  integer :: rads_flag              = 1       !< radar reflectivity for snow
+                                              !< 1: Mark Stoelinga (2005)
+                                              !< 2: Smith et al. (1975), Tong and Xue (2005)
+                                              !< 3: Marshall-Palmer formula (https://en.wikipedia.org/wiki/DBZ_(meteorology))
```
```suggestion
                                              !< 3: Marshall-Palmer formula (https://glossary.ametsoc.org/wiki/Marshall-palmer_relation)
```

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  integer :: radg_flag              = 1       !< radar reflectivity for graupel
+                                              !< 1: Mark Stoelinga (2005)
+                                              !< 2: Smith et al. (1975), Tong and Xue (2005)
+                                              !< 3: Marshall-Palmer formula (https://en.wikipedia.org/wiki/DBZ_(meteorology))
```
```suggestion
                                              !< 3: Marshall-Palmer formula (https://glossary.ametsoc.org/wiki/Marshall-palmer_relation)
```

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

grantfirl commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+  ! GFDL MP common (v1/v3) parameters, with different default values
+  ! #####################################################################################
+#ifdef GFDLMP_V3
+  real(kind_phys) :: tice           = 273.15  !< freezing temperature (K): ref: GFDL, GFS (DJS: V3=273.15)
```
Why does this need to be redefined from the one that is already provided by the host (temperature_at_zero_celsius)? Does it really make that much of a difference to 1) use a different value for each version and 2) use something different than the host and rest of physics?

dustinswales replied:
@grantfirl These are the defaults and (potentially) overwritten later on. I don't know if the 0.01K difference is by design or an oversight (I didn't want to add any changes when creating this module).
@RuiyuSun Does v3 need tice=273.16? or could we use 273.15?

RuiyuSun replied:
@dustinswales In v3 it is defined as 273.15.

RuiyuSun replied:
According to the comment in the v3 code, 273.15 is used to be consistent with the value defined in other schemes in GFS

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3.F90:
```diff
@@ -0,0 +1,392 @@
+
+      implicit none
+
+      ! DH* TODO: CLEANUP, all of these should be coming in through the argument list
```
Yes, and consider the comment RE: temperature_at_zero_celsius

dustinswales replied:
done.

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3.F90:
```diff
@@ -0,0 +1,392 @@
+        do k=1,levs
+          kk = levs-k+1
+          do i=1,im
+            if (imp_physics==imp_physics_gfdl) then 
```
Why is this necessary? Shouldn't the check above already have been executed and returned with an error if imp_physics is wrong?

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+!***********************************************************************
+!*                   GNU Lesser General Public License
+!*
+!* This file is part of the FV3 dynamical core.
```
It's a little weird to have this in a physics scheme in the CCPP, but I'm guessing that it's not our call.

dustinswales replied:
My guess is this is a relic from a prior code port. I guess we can leave it for now

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+    ! physics constants
+    ! -----------------------------------------------------------------------
+
+    real, parameter :: grav = 9.80665 ! acceleration due to gravity (m/s^2), ref: IFS
```
Yikes. Lots of redefining constants here. Why does GFDL MP need to have a different gravity or pi than the host or rest of physics?

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+    real, parameter :: eps = rdgas / rvgas ! 0.6219934994582882
+    real, parameter :: epsm1 = rdgas / rvgas - 1. ! -0.3780065005417118
+
+    real, parameter :: tice = 273.15 ! freezing temperature (K): ref: GFDL, GFS
```
This is double redefining. It was redefined in the GFDL MP specific module then redefined here. Why?

dustinswales replied:
Fixed.

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+    ! major cloud microphysics driver
+    ! -----------------------------------------------------------------------
+
+    call mpdrv (hydrostatic, ua, va, wa, delp, pt, qv, ql, qr, qi, qs, qg, qa, &
```
This is another scheme with 3 layers of drivers. Not ideal, but I assume that there is a good reason to keep all of them?

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+    ! unit convert to mm/day
+    ! -----------------------------------------------------------------------
+
+    convt = 86400. * rgrav / dts
```
redefining # of seconds in a day

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+        if (consv_checker) then
+            if (abs (sum (te_end_d (i, :)) + te_b_end_d (i) - sum (te_beg_d (i, :)) - te_b_beg_d (i)) / &
+                 (sum (te_beg_d (i, :)) + te_b_beg_d (i)) .gt. te_err) then
+                print*, "GFDL-MP-DRY TE: ", &
```
Consider wrapping print statements with a debug flag to reduce standard out spam for normal execution.

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+        case ("qg")
+            q = qg
+        case default
+            print *, "gfdl_mp: qflag error!"
```
Should one pass down and set ccpp_err_msg and ccpp_err_flag and return? Or is this error not that egregious?

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+                        case ("qg")
+                            qg (k) = q (k)
+                        case default
+                            print *, "gfdl_mp: qflag error!"
```
Same question RE: error handling

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+        case ("qg")
+            q = qg
+        case default
+            print *, "gfdl_mp: qflag error!"
```
Same question RE: error handling

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+        case ("qg")
+            q = qg
+        case default
+            print *, "gfdl_mp: qflag error!"
```
Same question RE: error handling

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+        case ("qg")
+            qg = q
+        case default
+            print *, "gfdl_mp: qflag error!"
```
Ditto

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+                        a4 (4, n) * r3 * (pr * (pr + pl) + pl ** 2)
+                    qm (k) = qm (k) * (ze (k) - ze (k + 1))
+                    k0 = n
+                    goto 555
```
GOTOs are generally frowned upon for CCPP schemes. Consider trying to use a different control structure if possible.

grantfirl commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3_mod.F90:
```diff
@@ -0,0 +1,7377 @@
+         te (k) = rgrav * te (k) * delp (k)
+         tw (k) = rgrav * (qv (k) + q_cond) * delp (k)
+     enddo
+     te_b = (dte + (lv00 * c_air * vapor - li00 * c_air * (ice + snow + graupel)) * dts / 86400 + sen * dts + stress * dts)
```
redefining # of seconds in a day again

grantfirl commented:
@RuiyuSun @dustinswales I finished my initial review. Mostly minor comments. The biggest things for me are that there are a lot of physical constants being redefined that are introducing potential inconsistencies with the rest of the code base (host and rest of physics) and the potential need to extend CCPP error handling downward into the actual physics code to catch and return errors. Overall, nice job! I don't see anything that absolutely HAS to be changed, although addressing my comments would definitely help CCPP compatibility.

grantfirl reviewed (changes requested):
I'm putting a "request changes" review on this. I'm not saying that all my comments need to lead to code changes -- explanations for why they can't change can also suffice, and I'll approve at that time.

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Address reviewers comments

dustinswales commented:
@grantfirl I changed the constants in the ccpp driver to inout arguments. The constants in the scheme module I changed to use the values in physcons. Adding all the "plumbing" to pass the constants around in the scheme is a biggish task, so I went with this. 
I also parameterized the precision in the scheme module.
There is probably more that could be done, but not in this initial PR.

grantfirl reviewed (approved):
I don't love bringing in physcons which means that this scheme can't be used outside of the UFS ecosystem, but if you just want to get this in and worry about it later, that is OK, I guess.

AndersJensen-NOAA committed:
Initial TEMPO commit to CCPP

AndersJensen-NOAA committed:
Changes for TEMPO to be called with imp_physics option 88

AndersJensen-NOAA committed:
Support for TEMPO aerosol-aware in CCPP

AndersJensen-NOAA committed:
Couple TEMPO to radiation

AndersJensen-NOAA committed:
Bugfix to radiation call TEMPO

AndersJensen-NOAA committed:
Add support for TEMPO hail-aware option

mkavulich commented on physics/MP/GFDL/module_gfdlmp_param.F90:
```diff
@@ -0,0 +1,413 @@
+        present(errmsg)) then
+
+       if ((version .ne. 1) .and. (version .ne. 3)) then
+          write (6, *) 'gfdl - mp :: invalid scheme version number'
```
```suggestion
          write (6, *) 'gfdl - mp :: invalid scheme version number', version
```
I always like error messages to echo the invalid parameter

mkavulich commented on physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3.F90:
```diff
@@ -0,0 +1,362 @@
+      enddo
+
+      ! reset precipitation amounts to zero
+      water0     = 0
```
```suggestion
      water0    = 0
```

mkavulich reviewed (commented):
@dustinswales I started this review a long time ago and never found time to finish it. These are not urgent corrections, so I'll leave them as optional. I didn't want to mark it as "approval" since I didn't finish reviewing the code, but what I got through looked good.

AndersJensen-NOAA committed:
Add hail-aware variables

AndersJensen-NOAA committed:
rename module variables

AndersJensen-NOAA committed:
Add branch to gitmodules file

AndersJensen-NOAA committed:
update TEMPO pointer to authoritative

grantfirl commented:
@dustinswales I'm not sure whether you want to address some of the changes for @mkavulich or go ahead with this PR as-is. Please let me know.

dustinswales committed:
GFDL MP developer (Linjiong Zhou) provided the code changes. The fix solved the problem of overly produced precip and sensitivity to microphysics timestep

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales commented:
@grantfirl @mkavulich This is ready to go as is.
There is more that could be done, but I can't take it on at the moment.

grantfirl commented:
> @grantfirl @mkavulich This is ready to go as is. There is more that could be done, but I can't take it on at the moment.

OK, I'll try to get this into the queue.

grantfirl committed:
Merge branch 'ufs/dev' into feature/gfdlmpv3

RuiyuSun reviewed (approved):
Thanks@dustinswales

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'feature/gfdlmpv3' of https://github.com/dustinswales/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Bugfix for name conflict.

dustinswales committed:
Add import reassignments to avoid error

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

grantfirl committed:
Merge branch 'ufs/dev' into tempo_dev

grantfirl committed:
set TEMPO constants from host

grantfirl committed:
Merge branch 'ufs/dev' into tempo_dev

grantfirl committed:
Merge branch 'ufs/dev' into tempo_dev_gjf

grantfirl committed:
Merge branch 'tempo_dev_gjf' into tempo_dev

grantfirl committed:
add re-initialization lines for am_i and am_g in mp_tempo_init (forgotten last commit)

grantfirl committed:
update TEMPO to latest main branch

grantfirl committed:
Merge branch 'ufs/dev' into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'feature/gfdlmpv3' of https://github.com/dustinswales/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

grantfirl committed:
Merge branch 'ufs/dev' into tempo_dev

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

Commit:
Remove PP directive

grantfirl committed:
Merge branch 'ufs/dev' into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into feature/gfdlmpv3

dustinswales committed:
Merge branch 'ufs/dev' of https://github.com/ufs-community/ccpp-physics into HEAD

dustinswales committed:
Update TEMPO submodule hash

dustinswales committed:
Merge branch 'tempo_dev' of https://github.com/AndersJensen-NOAA/ccpp-physics into feature/gfdlmpv3

grantfirl committed:
Merge branch 'ufs/dev' into feature/gfdlmpv3
